File: src/snippets/search/binary-search/binary-search.js
/**
 * Binary Search Function
 * @param {Array} sortedArray - The array to search through. Must be sorted.
 * @param {number} target - The value to search for.
 * @returns {number} - The index of the target value in the array, or -1 if not found.
 */
function binarySearch(sortedArray, target) {
    let left = 0;
    let right = sortedArray.length - 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);

        if (sortedArray[mid] === target) {
            return mid;
        } else if (sortedArray[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return -1; // Target not found
}

// Example usage:
const array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
const target = 5;
const result = binarySearch(array, target);

console.log(`Target ${target} found at index: ${result}`); // Output: Target 5 found at index: 4

File: src/snippets/search/binary-search/binary-search.lua
-- Binary Search Function
function binarySearch(arr, target)
    local low = 1
    local high = #arr

    while low <= high do
        local mid = math.floor((low + high) / 2)
        if arr[mid] == target then
            return mid -- Target found, return its index
        elseif arr[mid] < target then
            low = mid + 1 -- Search in the right half
        else
            high = mid - 1 -- Search in the left half
        end
    end

    return -1 -- Target not found
end

-- Example usage
local sortedArray = {1, 3, 5, 7, 9, 11, 13, 15, 17, 19}
local target = 7
local result = binarySearch(sortedArray, target)

if result ~= -1 then
    print("Target found at index: " .. result)
else
    print("Target not found")
end